File: src/GUI/objects/TableCell/TextStarRatingCellFactory.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package GUI.objects.TableCell;

import AudioPlayer.tagging.Metadata;
import static java.lang.Math.round;
import javafx.geometry.Pos;
import javafx.scene.control.TableCell;
import javafx.scene.control.TableColumn;
import util.parsing.StringParseStrategy;
import static util.parsing.StringParseStrategy.From.CONSTRUCTOR;
import static util.parsing.StringParseStrategy.To.CONSTANT;

/** Cell for rating displaying the value as text from '' to '*****'. */
@StringParseStrategy( from = CONSTRUCTOR, to = CONSTANT, constant = "Text star" )
public class TextStarRatingCellFactory implements RatingCellFactory {
    private static final String s0 = "";
    private static final String s1 = "*";
    private static final String s2 = "**";
    private static final String s3 = "***";
    private static final String s4 = "****";
    private static final String s5 = "*****";

    @Override
    public TableCell<Metadata, Double> call(TableColumn<Metadata, Double> param) {
        return new TableCell<Metadata,Double>(){
            {
                setAlignment(Pos.CENTER);
            }
            @Override
            protected void updateItem(Double item, boolean empty) {
                super.updateItem(item, empty);
                if(empty) {
                    setText(null);
                } else {
                    int r = (int)round(item/0.2);
                    String s;
                    if(r==0) s = s0;
                    else if(r==1) s = s1;
                    else if(r==2) s = s2;
                    else if(r==3) s = s3;
                    else if(r==4) s = s4;
                    else if(r==5) s = s5;
                    else s = null;
                    setText(s);
                }
            }
        };
    }
}

File: src/player/main/sp/it/pl/gui/objects/tablecell/TextStarRatingCellFactory.kt
package sp.it.pl.gui.objects.tablecell

import javafx.geometry.Pos
import javafx.scene.control.TableCell
import javafx.scene.control.TableColumn
import sp.it.pl.audio.tagging.Metadata
import sp.it.pl.util.parsing.StringParseStrategy
import sp.it.pl.util.parsing.StringParseStrategy.From
import sp.it.pl.util.parsing.StringParseStrategy.To
import java.lang.Math.round

/** Cell for rating displaying the value as text from '' to '*****'. */
@StringParseStrategy(from = From.SINGLETON, to = To.CONSTANT, constant = "Text star")
object TextStarRatingCellFactory: RatingCellFactory {
    private val s0 = ""
    private val s1 = "*"
    private val s2 = "**"
    private val s3 = "***"
    private val s4 = "****"
    private val s5 = "*****"

    override fun apply(param: TableColumn<Metadata, Double?>) = object: TableCell<Metadata, Double?>() {

        init {
            alignment = Pos.CENTER
        }

        override fun updateItem(item: Double?, empty: Boolean) {
            super.updateItem(item, empty)
            if (empty) {
                text = null
            } else {
                val r = round(item!!/0.2).toInt()
                text = when (r) {
                    0 -> s0
                    1 -> s1
                    2 -> s2
                    3 -> s3
                    4 -> s4
                    5 -> s5
                    else -> null
                }
            }
        }
    }

}